cmd/compact_unix.go

```diff
@@ -0,0 +1,199 @@
+//go:build !windows
+// +build !windows
+
+/*
+ * JuiceFS, Copyright 2020 Juicedata, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"fmt"
+	"github.com/juicedata/juicefs/pkg/chunk"
+	"github.com/juicedata/juicefs/pkg/meta"
+	"github.com/juicedata/juicefs/pkg/utils"
+	"github.com/juicedata/juicefs/pkg/vfs"
+	"github.com/urfave/cli/v2"
+	"sync"
+	"time"
+)
+
+func cmdCompact() *cli.Command {
+	return &cli.Command{
+		Name:      "compact",
+		Action:    compact,
+		Category:  "ADMIN",
+		Usage:     "Trigger compaction of slices",
+		ArgsUsage: "META-URL",
+		Description: `
+ Examples:
+ # compact with path
+ $ juicefs compact --path /mnt/jfs/foo redis://localhost
+
+ # max depth of 5
+ $ juicefs summary --path /mnt/jfs/foo --depth 5 /mnt/jfs/foo redis://localhost
+ `,
+		Flags: []cli.Flag{
+			&cli.StringSliceFlag{
+				Name:     "path",
+				Required: true,
+				Usage:    "path to be compact",
+			},
+			&cli.UintFlag{
+				Name:    "depth",
+				Aliases: []string{"d"},
+				Value:   2,
+				Usage:   "dir depth to be scan",
+			},
+			&cli.IntFlag{
+				Name:    "compact-concurrency",
+				Aliases: []string{"cc"},
+				Value:   10,
+				Usage:   "compact concurrency",
+			},
+			&cli.IntFlag{
+				Name:    "delete-concurrency",
+				Aliases: []string{"dc"},
+				Value:   10,
+				Usage:   "delete concurrency",
+			},
+		},
+	}
+}
+
+func compact(ctx *cli.Context) error {
+	setup(ctx, 1)
+	removePassword(ctx.Args().Get(0))
+
+	// parse flags
+	metaUri := ctx.Args().Get(0)
+	paths := ctx.StringSlice("path")
+
+	depth := ctx.Uint("depth")
+	if depth > 10 {
+		logger.Warn("depth should be <= 10")
+		depth = 10
+	}
+
+	deleteConcurrency := ctx.Int("delete-concurrency")
+	if deleteConcurrency <= 0 {
+		logger.Warn("thread number should be > 0")
+		deleteConcurrency = 1
+	}
+
+	compactConcurrency := ctx.Int("compact-concurrency")
+	if compactConcurrency <= 0 {
+		logger.Warn("thread number should be > 0")
+		compactConcurrency = 1
+	}
+
+	// new meta client
+	metaConf := meta.DefaultConf()
+	metaConf.MaxDeletes = int(deleteConcurrency)
+	client := meta.NewClient(metaUri, metaConf)
```

Comment:
Don't use meta client, instead write control message to a mount point. Check these codes:

- https://github.com/juicedata/juicefs/blob/7b85645cac16c8442b3b7d35365a99dc0e339cdd/cmd/info.go#L116
- https://github.com/juicedata/juicefs/blob/7b85645cac16c8442b3b7d35365a99dc0e339cdd/pkg/vfs/internal.go#L297